ompi/mca/mtl/psm2/mtl_psm2_component.c

```diff
@@ -77,9 +77,129 @@ mca_mtl_psm2_component_t mca_mtl_psm2_component = {
+    {MCA_BASE_VAR_TYPE_UNSIGNED_LONG, &ompi_mtl_psm2.psm2_mq_sendreqs_max, {.ulval = 1048576}, "PSM2_MQ_SENDREQS_MAX", OPAL_INFO_LVL_3,
+     "mq_sendreqs_max", "PSM2 maximum number of isend requests in flight (default: 1M)"},
+    {MCA_BASE_VAR_TYPE_UNSIGNED_LONG, &ompi_mtl_psm2.psm2_mq_recvreqs_max, {.ulval = 1048576}, "PSM2_MQ_RECVREQS_MAX", OPAL_INFO_LVL_3,
+     "mq_recvreqs_max", "PSM2 maximum number of irecv requests in flight (default: 1M)"},
+    {MCA_BASE_VAR_TYPE_UNSIGNED_LONG, &ompi_mtl_psm2.psm2_mq_rndv_hfi_threshold, {.ulval = 64000}, "PSM2_MQ_RNDV_HFI_THRESH", OPAL_INFO_LVL_3,
+     "hfi_eager_limit", "PSM2 eager to rendezvous threshold (default: 64000)"},
+    {MCA_BASE_VAR_TYPE_UNSIGNED_LONG, &ompi_mtl_psm2.psm2_mq_rndv_shm_threshold, {.ulval = 16000}, "PSM2_MQ_RNDV_SHM_THRESH", OPAL_INFO_LVL_3,
+     "shm_eager_limit", "PSM2 shared memory eager to rendezvous threshold (default: 16000)"},
+    {MCA_BASE_VAR_TYPE_BOOL, &ompi_mtl_psm2.psm2_recvthread, {.boolval = true}, "PSM2_RCVTHREAD", OPAL_INFO_LVL_3,
+     "use_receive_thread", "Use PSM2 progress thread (default: true)"},
+    {MCA_BASE_VAR_TYPE_BOOL, &ompi_mtl_psm2.psm2_shared_contexts, {.boolval = true}, "PSM2_SHAREDCONTEXTS", OPAL_INFO_LVL_6,
+     "use_shared_contexts", "Share PSM contexts between MPI processes (default: true)"},
+    {MCA_BASE_VAR_TYPE_UNSIGNED_LONG, &ompi_mtl_psm2.psm2_shared_contexts_max, {.ulval = 8}, "PSM2_SHAREDCONTEXTS_MAX", OPAL_INFO_LVL_9,
+     "max_shared_contexts", "Maximum number of contexts available on a node (default: 8, max: 8)"},
+    {MCA_BASE_VAR_TYPE_UNSIGNED_LONG, &ompi_mtl_psm2.psm2_tracemask, {.ulval = 1}, "PSM2_TRACEMASK", OPAL_INFO_LVL_9,
+     "trace_mask", "PSM2 tracemask value. See PSM2 documentation for accepted values (default: 1)"},
+    {-1},
+};
+
+static void ompi_mtl_psm2_set_shadow_env (struct ompi_mtl_psm2_shadow_variable *variable)
+{
+    mca_base_var_storage_t *storage = variable->storage;
+    char *env_value;
+    int ret;
+
+    switch (variable->variable_type) {
+    case MCA_BASE_VAR_TYPE_BOOL:
+        ret = asprintf (&env_value, "%s=%s", variable->env_name, storage->boolval ? "YES" : "NO");
+        break;
+    case MCA_BASE_VAR_TYPE_UNSIGNED_LONG:
+        if (0 == strcmp (variable->env_name, "PSM2_TRACEMASK")) {
+            /* PSM2 documentation shows the tracemask as a hexidecimal number. to be consitent
+             * use hexidecimal here. */
+            ret = asprintf (&env_value, "%s=0x%lx", variable->env_name, storage->ulval);
+        } else {
+            ret = asprintf (&env_value, "%s=%lu", variable->env_name, storage->ulval);
+        }
+        break;
+    case MCA_BASE_VAR_TYPE_STRING:
+        ret = asprintf (&env_value, "%s=%s", variable->env_name, storage->stringval);
+        break;
+    }
+
+    if (0 > ret) {
+        fprintf (stderr, "ERROR setting PSM2 environment variable: %s\n", variable->env_name);
+    }
+
+    putenv (env_value);
```

Comment:
Are you sure you always want to putenv, even if there was an error up in the asprintfs?